tools/filetop_getattr.py

```diff
@@ -0,0 +1,412 @@
+        find_path_levels_t = """
+            cur = cur->d_parent;
+            bpf_probe_read_kernel(&d_name, sizeof(d_name), (void *)&cur->d_name);
+            if (d_name.len > _PATH_PART_LEN_)
+                info.is_path_part_truncated = 1;
+            bpf_probe_read_kernel(&info.path{i}, sizeof(info.path{i}), d_name.name);
+        """
+
+        init_char = ""
+        find_path_levels = "struct dentry *cur = p->dentry;\n"
+
+        for i in range(self.args.path_levels):
+            init_char += f"char path{i}[{self.args.path_part_length}];\n"
+            find_path_levels += find_path_levels_t.format(i=i)
+
+        bpf_text = bpf_text.replace("_PATH_LEVEL_CHAR_DEF_", init_char)
+        bpf_text = bpf_text.replace("_FIND_PATH_BLOCK_", find_path_levels)
+        bpf_text = bpf_text.replace(
+            "_PATH_FILENAME_LEN_", str(self.args.path_filename_length)
+        )
+        bpf_text = bpf_text.replace("_PATH_PART_LEN_", str(self.args.path_part_length))
+
+        # code generation for filtering by comm
+        filter_comm_t = """
+            // Filter length is larger than what we're comparing with, so this
+            // is definitely a mismatch. This condition can technically be
+            // evaluated at the userspace when the program is first executed,
+            // but we don't have access to TASK_COMM_LEN. That said, if we hit
+            // this condition, we get 0 results, and the user should know why.
+            if (_TARGET_COMM_SIZE_VAR_ >= TASK_COMM_LEN) {
+                return 0;
+            }
+            // define local comm[] here to avoid initializing "struct info"
+            // if this operation is filtered out
+            char comm[TASK_COMM_LEN] = {0};
+            bpf_get_current_comm(&comm, sizeof(comm));
+            char target_comm[] = "_TARGET_COMM_VAR_";
+            for (int i = 0; i < _TARGET_COMM_SIZE_VAR_; ++i) {
+                // comm is shorter, so it has to be a mismatch.
+                if (comm[i] == '\\0' || target_comm[i] != comm[i]) {
+                    return 0;
+                }
+            }
+            // handle the case where comm is longer than target_comm. Note that
+            // target_comm[_TARGET_COMM_SIZE_VAR_] is \\0.
+            if (comm[_TARGET_COMM_SIZE_VAR_] != '\\0') {
+                return 0;
+            }
+        """
+        filter_comm = ""
+        if self.args.comm:
+            filter_comm = filter_comm_t.replace("_TARGET_COMM_VAR_", self.args.comm)
+            filter_comm = filter_comm.replace(
+                "_TARGET_COMM_SIZE_VAR_",
+                str(len(self.args.comm)),
+            )
+        bpf_text = bpf_text.replace("_FILTER_COMM_BLOCK_", filter_comm)
+
+        # dry run option
+        if self.args.dry_run:
+            print(bpf_text)
+            sys.exit()
+
+        # initialize BPF
+        self.bpf = bcc.BPF(text=bpf_text)
+        self.bpf.attach_kprobe(event="vfs_getattr", fn_name="trace_getattr_entry")
```

Comment:
Here, we attached to vfs_getattr(), which is tracing not only nfs but also stat() syscall etc. Is this really what you want? I thought you try to check and debug nfs layer. Is it possible to kprobe some nfs related functions?

Reply:
Yes - took a similar idea from the original `filetop` script here, but we added the `-c` option which filters the process name, so when we do `-c nfsd` it captures NFSd-related calls while the tool itself is still generic.